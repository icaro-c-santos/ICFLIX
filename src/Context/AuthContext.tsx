import { createContext, useEffect, useState } from "react";

export type TAuthData = {
  login?: string;
  token?: string;
  refreshToken?: string;
  name?: string;
  avatarUrl?: string;
  isLoggedIn?: boolean;
};

export type TAuthContext = {
  userLogged: TAuthData;
  setUserLogged: (userLogged: TAuthData) => void;
};

export const AuthContext = createContext<TAuthContext>({} as TAuthContext);

export const AuthProvider = (props: { children: any }) => {
  const [userLogged, setUserLogged] = useState<TAuthData>({});

  useEffect(() => {
    const data = localStorage.getItem("userLogged");
    if (data) {
      const user: TAuthData = JSON.parse(data);
      user.isLoggedIn && setUserLogged(user);
    } else {
      setUserLogged({});
    }
  }, []);

  return (
    <AuthContext.Provider value={{ userLogged, setUserLogged } as TAuthContext}>
      {props.children}
    </AuthContext.Provider>
  );
};
