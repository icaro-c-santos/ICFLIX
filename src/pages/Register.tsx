import { Box, Button, TextField } from "@mui/material";
import { useEffect } from "react";
import { useState } from "react";
import { useNavigate } from "react-router";
import {
  validatorLogin,
  validatorName,
  validatorPassword,
} from "../utils/validator";
import { Fragment } from "react";
import { bancoFake, serviceAuthentic } from "../services/User";
import ModalSucess from "../Components/ModalSucess/ModalSucess";
import ModalError from "../Components/ModalError/ModalError";

export const Register = () => {
  const [login, setLogin] = useState("");
  const [name, setName] = useState("");
  const [password, setPassowrd] = useState("");
  const [passwordConfirm, setPassowrdConfirm] = useState("");
  const [errorTextLogin, setErrorTextLogin] = useState<string[]>([]);
  const [errorTextName, setErrorTextName] = useState<string[]>([]);
  const [errorTextPassword, setErrorTextPassword] = useState<string[]>([]);
  const [errorTextPasswordConfirm, setErrorTextPasswordConfirm] = useState("");
  const [messageError, setMessageError] = useState("");
  const [openModalError, setOpenModalError] = useState(false);
  const [openModalSucess, setOpenModalSucess] = useState(false);
  const [messageSucess, setMessageSucess] = useState("");
  const navigate = useNavigate();

  const closeModalSucess = (value: boolean) => {
    setOpenModalSucess(value);
    navigate("/login");
  };

  const handlerLogin = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    setLogin(event.target.value);
  };

  const handlerName = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    setName(event.target.value);
  };

  const handlerPassword = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement> | null
  ) => {
    if (event == null) {
      return;
    }
    setPassowrd(event.target.value);
  };

  const handlerPasswordConfirm = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    setPassowrdConfirm(event.target.value);
  };

  useEffect(() => {
    console.log("useEffect login");
    const errosLogin = validatorLogin(login);
    if (
      (login.length <= 0 &&
        name.length <= 0 &&
        password.length <= 0 &&
        passwordConfirm.length <= 0) ||
      errosLogin.length <= 0
    ) {
      setErrorTextLogin([]);
    } else {
      setErrorTextLogin(errosLogin);
    }

    const errosName = validatorName(name);
    if (
      (name.length <= 0 &&
        password.length <= 0 &&
        passwordConfirm.length <= 0) ||
      errosName.length <= 0
    ) {
      setErrorTextName([]);
    } else {
      setErrorTextName(errosName);
    }

    if (passwordConfirm.length > 0 && password != passwordConfirm) {
      setErrorTextPasswordConfirm("AS SENHAS DEVEM SER IGUAIS");
    } else {
      setErrorTextPasswordConfirm("");
    }

    const errosPassword = validatorPassword(password);
    if (password.length <= 0 || errosPassword.length <= 0) {
      setErrorTextPassword([]);
    } else {
      setErrorTextPassword(errosPassword);
    }
  }, [login, name, password, passwordConfirm]);

  const handlerButtonRegister = async () => {
    let erros: string[] = [];
    erros = erros.concat(
      validatorLogin(login),
      validatorName(name),
      validatorPassword(password)
    );
    if (erros.length <= 0 && password == passwordConfirm) {
      try {
        console.log("a",bancoFake)
        await serviceAuthentic.registerUser({
          login: login,
          name: name,
          password: password,
        });
        setMessageSucess("USUARIO CADASTRADO COM SUCESSO!");
        setOpenModalSucess(true);
      } catch (error) {
        setMessageError(error.message);
        setOpenModalError(true);
      }
    } else {
      setMessageError("PREENCHA TODOS OS CAMPOS CORRETAMENTE!");
      setOpenModalError(true);
    }
  };

  return (
    <>
      <ModalSucess
        message={messageSucess}
        openModal={openModalSucess}
        setOpenModal={closeModalSucess}
      />
      <ModalError
        message={messageError}
        openModal={openModalError}
        setOpenModal={setOpenModalError}
      />
      <Box
        sx={{
          display: "flex",
          flexDirection: "column",
          maxWidth: "400px",
          margin: "50px auto",
          justifyContent: "center",
          bgcolor: "#e6e6e6",
          padding: "40px",
          borderRadius: "25px",
          gap: "20px",
          boxShadow: "0px 2px 13px 0px #0000008f",
        }}
      >
        <Box
          sx={{
            gap: "20px",
            justifyContent: "center",
            display: "flex",
            flexDirection: "column",
          }}
        >
          <TextField
            placeholder="Login"
            onChange={handlerLogin}
            value={login}
            error={errorTextLogin.length > 0}
            helperText={errorTextLogin.map((error) => (
              <Fragment key={Math.random()}>{error}</Fragment>
            ))}
          ></TextField>
          <TextField
            placeholder="Nome"
            onChange={handlerName}
            value={name}
            error={errorTextName.length > 0}
            helperText={errorTextName}
          ></TextField>
          <TextField
            placeholder="Senha"
            onChange={handlerPassword}
            value={password}
            type={"password"}
            error={errorTextPassword.length > 0}
            helperText={errorTextPassword.map((error) => (
              <Fragment key={Math.random()}>{error}</Fragment>
            ))}
          ></TextField>
          <TextField
            placeholder="Confirmar Senha"
            onChange={handlerPasswordConfirm}
            value={passwordConfirm}
            type={"password"}
            error={errorTextPasswordConfirm.length > 0}
            helperText={errorTextPasswordConfirm}
          ></TextField>
        </Box>
        <Box sx={{ display: "flex", justifyContent: "space-between" }}>
          <Button
            variant="outlined"
            sx={{ padding: "10px", borderRadius: "10px", width: "100%" }}
            onClick={handlerButtonRegister}
          >
            Registrar
          </Button>
        </Box>
      </Box>
    </>
  );
};
