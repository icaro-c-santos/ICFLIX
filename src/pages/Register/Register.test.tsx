import { render, screen, within } from "@testing-library/react";
import { AuthProvider } from "../../Context/AuthContext";
import userEvent from "@testing-library/user-event";
import { act } from "react-dom/test-utils";
import { waitFor } from "@testing-library/react";
import { Register } from "./Register";
import clientAuth from "../../Client/User";
import renderer from "react-test-renderer";
const mockNavigate = jest.fn();
jest.mock("react-router", () => ({
  ...jest.requireActual("react-router"),
  useNavigate: () => mockNavigate,
}));

const renderRegister = () => {
  render(
    <AuthProvider>
      <Register />
    </AuthProvider>
  );
};

const renderRegisterAct = async () => {
  await act(() => {
    render(
      <AuthProvider>
        <Register />
      </AuthProvider>
    );
  });
};

describe("<Register>", () => {
  it("should render Correctly", () => {
    renderRegister();

    expect(screen.getByPlaceholderText(/login/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/nome/i)).toBeInTheDocument();
    expect(screen.getByTestId(/input-password/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/confirmar senha/i)).toBeInTheDocument();
    expect(
      screen.getByRole("button", {
        name: /registrar/i,
      })
    ).toBeInTheDocument();
  });

  it("should render (error) when not insert somethings in field confirmar senha", async () => {
    renderRegister();
    await waitFor(() => {
      userEvent.type(screen.getByPlaceholderText(/confirmar senha/i), "12345");
    });

    expect(
      screen.getByText(/login não pode ficar em branco/i)
    ).toBeInTheDocument();
    expect(
      screen.getByText(/nome não pode ficar em branco/i)
    ).toBeInTheDocument();
    expect(screen.getByText(/as senhas devem ser iguais/i)).toBeInTheDocument();
  });

  it("should render (error) when insert somethings in field login", async () => {
    await renderRegisterAct();
    await waitFor(() => {
      const view = screen.getByTestId("input-password");
      userEvent.type(within(view).getByPlaceholderText(/senha/i), "1");
    });
    await waitFor(() => {
      expect(
        screen.getByText(
          /pelo menos 8 caracters\.pelo menos 2 numeros\.pelo menos 1 caractere especial: !\-@\-#\-\$\-%\-&\-\?\-\*/i
        )
      ).toBeInTheDocument();
    });
  });

  it("should render (error) when insert somethings in field login", async () => {
    await renderRegisterAct();
    await waitFor(() => {
      const view = screen.getByTestId("input-password");
      userEvent.type(within(view).getByPlaceholderText(/senha/i), "12");
    });
    await waitFor(() => {
      expect(
        screen.getByText(
          /pelo menos 8 caracters\.pelo menos 1 caractere especial: !\-@\-#\-\$\-%\-&\-\?\-\*/i
        )
      ).toBeInTheDocument();
    });
  });

  it("should render error (USUARIO JÁ EXISTENTE!)", async () => {
    await renderRegisterAct();
    await waitFor(() => {
      const view = screen.getByTestId("input-password");
      userEvent.type(within(view).getByPlaceholderText(/senha/i), "12#");
    });
    await waitFor(() => {
      expect(screen.getByText(/pelo menos 8 caracters\./i)).toBeInTheDocument();
    });
  });

  it("should render (error ?) when insert somethings in field login", async () => {
    const mockClienteAuth = jest
      .spyOn(clientAuth, "registerUser")
      .mockRejectedValue(new Error("USUARIO JÁ EXISTENTE!"));

    await renderRegisterAct();
    await waitFor(() => {
      const view = screen.getByTestId("input-password");
      userEvent.type(within(view).getByPlaceholderText(/senha/i), "12#Aa2345");
      userEvent.type(
        screen.getByPlaceholderText(/login/i),
        "icarotest@test.com"
      );
      userEvent.type(
        screen.getByPlaceholderText(/nome/i),
        "icarotest@test.com"
      );
      userEvent.type(
        screen.getByPlaceholderText(/confirmar senha/i),
        "12#Aa2345"
      );
    });
    await waitFor(() => {
      userEvent.click(screen.getByRole("button", { name: /registrar/i }));
    });

    await waitFor(() => {
      expect(
        screen.getByRole("dialog", { name: /USUARIO JÁ EXISTENTE!/i })
      ).toBeInTheDocument();
    });
  });

  it("should render (USUARIO CADASTRADO COM SUCESSO!)", async () => {
    jest.spyOn(clientAuth, "registerUser").mockResolvedValue({} as any);

    await renderRegisterAct();
    await waitFor(() => {
      const view = screen.getByTestId("input-password");
      userEvent.type(within(view).getByPlaceholderText(/senha/i), "12#Aa2345");
      userEvent.type(
        screen.getByPlaceholderText(/login/i),
        "icarotest@test.com"
      );
      userEvent.type(
        screen.getByPlaceholderText(/nome/i),
        "icarotest@test.com"
      );
      userEvent.type(
        screen.getByPlaceholderText(/confirmar senha/i),
        "12#Aa2345"
      );
    });
    await waitFor(() => {
      userEvent.click(screen.getByRole("button", { name: /registrar/i }));
    });

    await waitFor(() => {
      expect(
        screen.getByRole("dialog", { name: /USUARIO CADASTRADO COM SUCESSO!/i })
      ).toBeInTheDocument();
    });
  });

  it("Matches DOM Snapshot", () => {
    const domTree = renderer
      .create(
        <AuthProvider>
          <Register />
        </AuthProvider>
      )
      .toJSON();
    expect(domTree).toMatchSnapshot();
  });
});
