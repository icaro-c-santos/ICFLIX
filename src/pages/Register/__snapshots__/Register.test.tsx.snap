// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Register> Matches DOM Snapshot 1`] = `
Array [
  <div />,
  <div />,
  <div
    className="MuiBox-root css-hhdv6n"
  >
    <div
      className="MuiBox-root css-1yrfh49"
    >
      <div
        className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-describedby=":r2:-helper-text"
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id=":r2:"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Login"
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-ihdtdm"
            >
              <span
                className="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          className="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
          id=":r2:-helper-text"
        />
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-describedby=":r3:-helper-text"
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id=":r3:"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Nome"
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-ihdtdm"
            >
              <span
                className="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          className="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
          id=":r3:-helper-text"
        />
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        data-testid="input-password"
        datatype="password"
      >
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-describedby=":r4:-helper-text"
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id=":r4:"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Senha"
            required={false}
            type="password"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-ihdtdm"
            >
              <span
                className="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          className="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
          id=":r4:-helper-text"
        />
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id=":r5:"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Confirmar Senha"
            required={false}
            type="password"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-ihdtdm"
            >
              <span
                className="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      className="MuiBox-root css-gg4vpm"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-16k2pua-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Registrar
      </button>
    </div>
  </div>,
]
`;
